pyccel/ast/datatypes.py

```diff
@@ -12,115 +12,201 @@
-}
-iso_c_binding_shortcut_mapping = {
-    'C_INT8_T'              : 'i8',
-    'C_INT16_T'             : 'i16',
-    'C_INT32_T'             : 'i32',
-    'C_INT64_T'             : 'i64',
-    'C_INT128_T'            : 'i128',
-    'C_FLOAT'               : 'f32',
-    'C_DOUBLE'              : 'f64',
-    'C_LONG_DOUBLE'         : 'f128',
-    'C_FLOAT_COMPLEX'       : 'c32',
-    'C_DOUBLE_COMPLEX'      : 'c64',
-    'C_LONG_DOUBLE_COMPLEX' : 'c128',
-    'C_BOOL'                : 'b1'
-}
+class PrimitiveType(metaclass=Singleton):
+    """
+    Base class representing types of datatypes.
+
+    The base class representing the category of datatype to which a FixedSizeType
+    may belong (e.g. integer, floating point).
+    """
+    __slots__ = ()
+    _name = '__UNDEFINED__'
+
+    def __init__(self): #pylint: disable=useless-parent-delegation
+        # This __init__ function is required so the ArgumentSingleton can
+        # always detect a signature
+        super().__init__()
+
+    def __str__(self):
+        return self._name
+
+    def __reduce__(self):
+        """
+        Function called during pickling.
+
+        For more details see : https://docs.python.org/3/library/pickle.html#object.__reduce__.
+        This function is necessary to ensure that DataTypes remain singletons.
+
+        Returns
+        -------
+        callable
+            A callable to create the object.
+        args
+            A tuple containing any arguments to be passed to the callable.
+        """
+        return (self.__class__, ())
+
+class PyccelBooleanType(PrimitiveType):
+    """
+    Class representing a boolean datatype.
+
+    Class representing a boolean datatype.
+    """
+    __slots__ = ()
+    _name = 'boolean'
+
+class PyccelIntegerType(PrimitiveType):
+    """
+    Class representing an integer datatype.
+
+    Class representing an integer datatype.
+    """
+    __slots__ = ()
+    _name = 'integer'
+
+class PyccelFloatingPointType(PrimitiveType):
+    """
+    Class representing a floating point datatype.
+
+    Class representing a floating point datatype.
+    """
+    __slots__ = ()
+    _name = 'floating point'
+
+class PyccelComplexType(PrimitiveType):
+    """
+    Class representing a complex datatype.
+
+    Class representing a complex datatype.
+    """
+    __slots__ = ()
+    _name = 'complex'
+
+class PyccelCharacterType(PrimitiveType):
+    """
+    Class representing a character datatype.
+
+    Class representing a character datatype.
+    """
+    __slots__ = ()
+    _name = 'character'
 
 #==============================================================================
 
-class DataType:
+class PyccelType:
     """
-    Base class representing native datatypes.
+    Base class representing the type of an object.
+
+    Base class representing the type of an object from which all
+    types must inherit. A type must contain enough information to
+    describe the declaration type in a low-level language.
```

Comment:
You say that all types must inherit from this class, but `PrimitiveType` is not a `PyccelType`.

And subclasses of `PrimitiveType` are `PyccelBooleanType`, `PyccelIntegerType`, `PyccelFloatingPointType`, etc... Given their names, one would expect these to be subclasses of `PyccelType`!

I think you might have mixed up some class names. For instance, `PythonNativeBool` is a subclass of `PyccelType`, and its `primitive_type` is `PyccelBooleanType` which is a `PrimitiveType`.

We need an inheritance class diagram!

Reply:
The confusion does not come from the word `Type`, but rather from the unclear use of the words `Python` and `Pyccel`.

Reply:
If `PyccelType` is a superclass, everyone (me included) expects `Pyccel[Boolean|Integer|...]Type` to be its subclass.

Reply:
Ah that also makes sense. The logic in my head was `PyccelType` is the type of the object inside Pyccel. `PythonNativeInt` is the type of the native integers found in Python. The word Python is important here to contrast with `CNativeInt`:
https://github.com/pyccel/pyccel/blob/790a1a3e8c5b268730cd17141f1d75f792898cde/pyccel/ast/c_concepts.py#L24-L32

A `PythonNativeInt` is an object with an integer primitive type and a precision specific to the platform (`int32_t` on windows, `int64_t` on linux/macos) while a `CNativeInt` is an object with an integer primitive type where the exact precision will not be printed in C (`int a;`).

Maybe we should rename the primitive sub-classes to clarify things. So `PyccelBooleanType` -> `PrimitiveBooleanType` ? Or simply `BooleanType` ? But I worry that the latter might be too general or might collide with other packages

Reply:
Fixed:
da4c15e05e65de1efd409ba94edeab4a98cdeb24
9d23308a1db784e0735e30a52df18f6f4134fc9f
856fbf9d741cf3f3750a25bcb1be40b3d744a7fd
ccb774245db7a62fb6444277c1f8972abc6575b3